stratum/hal/lib/barefoot/bf_sde_wrapper.h

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019-present Barefoot Networks, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_HAL_LIB_BAREFOOT_BF_SDE_WRAPPER_H_
+#define STRATUM_HAL_LIB_BAREFOOT_BF_SDE_WRAPPER_H_
+
+#include <memory>
+
+#include "absl/synchronization/mutex.h"
+#include "stratum/glue/integral_types.h"
+#include "stratum/glue/status/status.h"
+#include "stratum/glue/status/statusor.h"
+#include "stratum/hal/lib/barefoot/bf_sde_interface.h"
+#include "stratum/hal/lib/common/common.pb.h"
+#include "stratum/lib/channel/channel.h"
+
+namespace stratum {
+namespace hal {
+namespace barefoot {
+
+// The "BfSdeWrapper" is an implementation of BfSdeInterface which is used
+// on real hardware to talk to Tofino ASIC.
+class BfSdeWrapper : public BfSdeInterface {
+ public:
+  // Default MTU for ports on Tofino.
+  static constexpr int32 kBfDefaultMtu = 10 * 1024;  // 10K
+
+  // BfSdeInterface public methods.
+  ::util::StatusOr<PortState> GetPortState(int device, int port) override;
+  ::util::Status GetPortCounters(int device, int port,
+                                 PortCounters* counters) override;
+  ::util::Status RegisterPortStatusEventWriter(
+      std::unique_ptr<ChannelWriter<PortStatusEvent>> writer) override
+      LOCKS_EXCLUDED(port_status_event_writer_lock_);
+  ::util::Status UnregisterPortStatusEventWriter() override
+      LOCKS_EXCLUDED(port_status_event_writer_lock_);
+  ::util::Status AddPort(int device, int port, uint64 speed_bps,
+                         FecMode fec_mode) override;
+  ::util::Status DeletePort(int device, int port) override;
+  ::util::Status EnablePort(int device, int port) override;
+  ::util::Status DisablePort(int device, int port) override;
+  ::util::Status SetPortAutonegPolicy(int device, int port,
+                                      TriState autoneg) override;
+  ::util::Status SetPortMtu(int device, int port, int32 mtu) override;
+  bool IsValidPort(int device, int port) override;
+  ::util::Status SetPortLoopbackMode(int uint, int port,
+                                     LoopbackState loopback_mode) override;
+  ::util::StatusOr<uint32> GetPortIdFromPortKey(
+      int device, const PortKey& port_key) override;
+  ::util::StatusOr<int> GetPcieCpuPort(int device) override;
+  ::util::Status SetTmCpuPort(int device, int port) override;
+
+  // Creates the singleton instance. Expected to be called once to initialize
+  // the instance.
+  static BfSdeWrapper* CreateSingleton() LOCKS_EXCLUDED(init_lock_);
+
+  // The following public functions are specific to this class. They are to be
+  // called by SDE callbacks only.
+
+  // Return the singleton instance to be used in the SDE callbacks.
+  static BfSdeWrapper* GetSingleton() LOCKS_EXCLUDED(init_lock_);
+
+  // Called whenever a linkscan event is received from SDK. It forwards the
```

Comment:
linkscan

Reply:
yes we do. Found no other instance of `linkscan`.